strawberryfields/utils/post_processing.py

```diff
@@ -0,0 +1,341 @@
+
+    .. code-block:: python
+
+        samples = np.array([[2, 0],
+                            [2, 2],
+                            [2, 0],
+                            [0, 0]])
+
+    Getting the probabilities of all the Fock states based on these PNR samples:
+
+    >>> probs = all_fock_probs_pnr(samples)
+
+    2 out of 4 shots resulted in the ``(2, 0)`` state. We can check if the
+    probability for this state is correct by indexing into ``probs``:
+
+    >>> probs[(2, 0)]
+    0.5
+
+    We can further check the entire array of probabilities:
+    >>> probs
+    [[0.25 0.   0.  ]
+    [0.   0.   0.  ]
+    [0.5  0.   0.25]]
+
+    Args:
+        photon_number_samples (ndarray): the photon number samples with a shape
+            of (shots, modes)
+
+    Returns:
+        array: array of dimension :math:`\underbrace{D\times D\times D\cdots\times D}_{\text{num modes}}`
+            containing the Fock state probabilities, where :math:`D` is the Fock basis cutoff truncation
+    """
+    _check_samples(photon_number_samples)
+
+    num_modes = photon_number_samples.shape[1]
+
+    shots = photon_number_samples.shape[0]
+    num_modes = photon_number_samples.shape[1]
+    max_val = np.max(photon_number_samples)
+    bins = [list(range(max_val + 2)) for i in range(num_modes)]
+    H, _ = np.histogramdd(photon_number_samples, bins=bins)
+    probabilities = H / shots
+    return probabilities
+
+
+def _product_for_modes(samples, modes=None):
+    """Getting the product of samples across modes.
+
+    Args:
+        samples (ndarray): the photon number samples with a shape
+            of (shots, modes)
+        modes (Sequence): a flat sequence containing indices of modes to get
+            the expectation value for
+
+    Returns:
+        ndarray: product of the samples across modes
+    """
+    modes = np.array(modes)
+    selected_samples = samples[:, modes]
+    return np.prod(selected_samples, axis=1)
+
+
+def _check_samples(samples):
+    """Validation check for the input samples.
+
+    Checks include data types checks and dimension of the input samples.
+
+    Args:
+        samples (ndarray): the photon number samples with a shape of (shots,
+            modes)
+    """
+    if not isinstance(samples, np.ndarray) or samples.ndim != 2:
+        raise Exception("Samples needs to be represented as a two dimensional numpy array.")
+
+
+def _check_modes(samples, modes):
+    """Validation checks for the input modes.
+
+    Checks include data types checks, checks for the dimension of the inputs
+    and the validity of the modes specified.
+
+    Args:
+        samples (ndarray): the photon number samples with a shape
+            of (shots, modes)
+        modes (Sequence): the input modes to get the expectation value for, a
+            flattened sequence
+    """
+    num_modes = samples.shape[1]
+    flattened_sequence_indices_msg = (
+        "The input modes need to be specified as a flattened sequence of indices!"
```

Comment:
Should it be the following?
```suggestion
        "The input modes need to be specified as a sequence of positive integers."
```